source/fuzz/transformation_flatten_conditional_branch.cpp

```diff
@@ -0,0 +1,595 @@
+
+      if (instructions_to_fresh_ids[instruction].size() <
+          ids_needed_by_this_instruction) {
+        return false;
+      }
+    } else {
+      // If there is no mapping, we need to rely on the pool of
+      // overflow ids, where there must be enough remaining ids.
+
+      remaining_overflow_ids -= ids_needed_by_this_instruction;
+
+      if (remaining_overflow_ids < 0) {
+        return false;
+      }
+    }
+  }
+
+  // All checks were passed.
+  return true;
+}
+
+void TransformationFlattenConditionalBranch::Apply(
+    opt::IRContext* ir_context,
+    TransformationContext* transformation_context) const {
+  uint32_t header_block_id = message_.header_block_id();
+  auto header_block = ir_context->cfg()->block(header_block_id);
+
+  // Find the first block where flow converges (it is not necessarily the merge
+  // block).
+  uint32_t convergence_block_id = header_block->MergeBlockId();
+  while (ir_context->cfg()->preds(convergence_block_id).size() == 1) {
+    convergence_block_id = ir_context->cfg()->preds(convergence_block_id)[0];
+  }
+
+  // Get the mapping from instructions to fresh ids.
+  auto instructions_to_fresh_ids = GetInstructionsToFreshIdsMapping(ir_context);
+
+  // Keep track of the number of overflow ids used.
+  uint32_t overflow_ids_used = 0;
+
+  auto branch_instruction = header_block->terminator();
+
+  opt::BasicBlock* last_true_block = nullptr;
+
+  // Adjust the conditional branches by enclosing problematic instructions
+  // within conditionals and get references to the last block in each branch.
+  for (int branch = 2; branch >= 1; branch--) {
+    // branch = 1 corresponds to the true branch, branch = 2 corresponds to the
+    // false branch. Consider the false branch first so that the true branch is
+    // laid out right after the false branch.
+
+    auto block = header_block;
+    // Get the id of the first block in this branch.
+    uint32_t block_id = branch_instruction->GetSingleWordInOperand(branch);
+
+    // Consider all blocks in the branch until the convergence block is reached.
+    while (block_id != convergence_block_id) {
+      // Move the block to right after the previous one.
+      block->GetParent()->MoveBasicBlockToAfter(block_id, block);
+
+      block = ir_context->cfg()->block(block_id);
+      block_id = block->terminator()->GetSingleWordInOperand(0);
+
+      // Find all the problematic instructions in the block (OpStore, OpLoad,
+      // OpFunctionCall).
+      std::vector<opt::Instruction*> problematic_instructions;
+
+      block->ForEachInst(
+          [&problematic_instructions](opt::Instruction* instruction) {
+            switch (instruction->opcode()) {
+              case SpvOpStore:
+              case SpvOpLoad:
+              case SpvOpFunctionCall:
+                problematic_instructions.push_back(instruction);
+                break;
+              default:
+                break;
+            }
+          });
```

Comment:
Thanks for spotting this, I forgot to change it! I think using ``if (!InstructionHasNoSideEffects(...))`` would be fine, because we should be able to handle all the functions that return true when given to ``InstructionCanBeHandled`` and, if there is some instruction that cannot be handled, it will have already been spotted by IsApplicable.

Reply:
I have now updated this function